File: src/ghsingo/internal/audio/karplus.go
package audio

import "math/rand"

// KarplusVoice is a one-shot Karplus-Strong plucked-string voice tuned for
// short bell-like attacks. A fresh voice is allocated per trigger.
type KarplusVoice struct {
	buf       []float32
	pos       int
	decay     float64
	remaining int
	velocity  float32
}

// NewKarplusVoice starts a bell ring at the given frequency. decay should be
// 0.99~0.999 (higher = longer ring). velocity scales the initial excitation.
// Auto-silences after ~2 seconds regardless of decay (safety cap).
func NewKarplusVoice(sampleRate int, freq float64, decay float64, velocity float32) *KarplusVoice {
	n := int(float64(sampleRate) / freq)
	if n < 2 {
		n = 2
	}
	buf := make([]float32, n)
	for i := range buf {
		buf[i] = (rand.Float32()*2 - 1) * velocity
	}
	return &KarplusVoice{
		buf:       buf,
		decay:     decay,
		remaining: sampleRate * 2,
		velocity:  velocity,
	}
}

// NextSample returns one sample and advances the voice state.
func (v *KarplusVoice) NextSample() float32 {
	if v.remaining <= 0 {
		return 0
	}
	out := v.buf[v.pos]
	next := (v.pos + 1) % len(v.buf)
	averaged := (v.buf[v.pos] + v.buf[next]) * 0.5
	v.buf[v.pos] = float32(float64(averaged) * v.decay)
	v.pos = next
	v.remaining--
	return out
}

// Done reports whether the voice has stopped producing audio.
func (v *KarplusVoice) Done() bool {
	return v.remaining <= 0
}

File: src/rsghsing/src/audio/karplus.rs
//! Karplus-Strong one-shot bell voice. Ported from `internal/audio/karplus.go`.
//!
//! NOTE on determinism: Go seeds this voice from the *global* `math/rand`,
//! which Go 1.20+ auto-seeds per process — so two Go renders of the same
//! window are already not sample-identical. rsghsing seeds a local generator
//! from a fixed constant instead; the audio-parity tolerance (|Δ| ≤ 1.0 LU)
//! absorbs the difference and every render is reproducible.

use crate::gorand::GoRand;

pub struct KarplusVoice {
    buf: Vec<f32>,
    pos: usize,
    decay: f64,
    remaining: i32,
    velocity: f32,
}

impl KarplusVoice {
    /// `freq` in Hz, `decay` in 0.99..0.999, `velocity` scales excitation.
    /// Auto-silences after ~2 seconds (safety cap).
    pub fn new(sample_rate: i32, freq: f64, decay: f64, velocity: f32, seed: i64) -> Self {
        let mut n = (f64::from(sample_rate) / freq) as usize;
        if n < 2 {
            n = 2;
        }
        let mut rng = GoRand::new(seed);
        let buf: Vec<f32> = (0..n)
            .map(|_| (rng.float64() * 2.0 - 1.0) as f32 * velocity)
            .collect();
        KarplusVoice {
            buf,
            pos: 0,
            decay,
            remaining: sample_rate * 2,
            velocity,
        }
    }

    pub fn next_sample(&mut self) -> f32 {
        if self.remaining <= 0 {
            return 0.0;
        }
        let out = self.buf[self.pos];
        let next = (self.pos + 1) % self.buf.len();
        let averaged = (self.buf[self.pos] + self.buf[next]) * 0.5;
        self.buf[self.pos] = (f64::from(averaged) * self.decay) as f32;
        self.pos = next;
        self.remaining -= 1;
        out
    }

    pub fn done(&self) -> bool {
        self.remaining <= 0
    }
}
